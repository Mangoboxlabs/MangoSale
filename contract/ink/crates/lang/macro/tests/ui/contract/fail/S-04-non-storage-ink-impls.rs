use ink_lang as ink;

#[ink::contract]
mod non_storage_ink_impls {
    // This test ensures that ink! `impl` blocks are always
    // implemented on the only storage struct definition.

    #[ink(storage)]
    pub struct StorageStruct {}

    // This ink! `impl` block is okay.
    impl StorageStruct {
        #[ink(constructor)]
        pub fn constructor1() -> Self {
            Self {}
        }

        #[ink(message)]
        pub fn message1(&self) {}
    }

    // Missing the `#[ink(storage)]` attribute on purpose.
    pub struct NonStorageStruct {}

    // This ink! `impl` block is invalid in that it implements
    // the messages and constructors for a non-existing ink!
    // storage struct. We expect a failure here.
    impl NonStorageStruct {
        #[ink(constructor)]
        pub fn constructor2() -> Self {
            Self {}
        }

        #[ink(message)]
        pub fn message2(&self) {}
    }
}

fn main() {}
