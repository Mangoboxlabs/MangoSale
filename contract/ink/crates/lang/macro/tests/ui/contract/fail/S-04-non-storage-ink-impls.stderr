error[E0271]: type mismatch resolving `<NonStorageStruct as non_storage_ink_impls::_::_::{closure#0}::TypeEq>::This == StorageStruct`
  --> $DIR/S-04-non-storage-ink-impls.rs:28:10
   |
28 |     impl NonStorageStruct {
   |          ^^^^^^^^^^^^^^^^ expected struct `StorageStruct`, found struct `NonStorageStruct`
   |
note: required by a bound in `non_storage_ink_impls::_::_::{closure#0}::assert_type_eq_all`
  --> $DIR/S-04-non-storage-ink-impls.rs:28:10
   |
28 |     impl NonStorageStruct {
   |          ^^^^^^^^^^^^^^^^
   |          |
   |          required by a bound in this
   |          required by this bound in `non_storage_ink_impls::_::_::{closure#0}::assert_type_eq_all`
   = note: this error originates in the macro `::ink_lang::static_assertions::assert_type_eq_all` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `message2` found for struct `StorageStruct` in the current scope
  --> $DIR/S-04-non-storage-ink-impls.rs:35:16
   |
9  |     pub struct StorageStruct {}
   |     ------------------------ function or associated item `message2` not found for this
...
35 |         pub fn message2(&self) {}
   |                ^^^^^^^^
   |                |
   |                function or associated item not found in `StorageStruct`
   |                help: there is an associated function with a similar name: `message1`

error[E0599]: no function or associated item named `constructor2` found for struct `StorageStruct` in the current scope
  --> $DIR/S-04-non-storage-ink-impls.rs:30:16
   |
9  |     pub struct StorageStruct {}
   |     ------------------------ function or associated item `constructor2` not found for this
...
30 |         pub fn constructor2() -> Self {
   |                ^^^^^^^^^^^^
   |                |
   |                function or associated item not found in `StorageStruct`
   |                help: there is an associated function with a similar name: `constructor1`
